pkg/controller/certificatesigningrequest/approver_controller.go

```diff
@@ -0,0 +1,335 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	sautil "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	csrinformer "k8s.io/client-go/informers/certificates/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	csrlister "k8s.io/client-go/listers/certificates/v1"
+	"k8s.io/client-go/tools/cache"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	config "antrea.io/antrea/pkg/config/controller"
+	"antrea.io/antrea/pkg/util/env"
+)
+
+const (
+	approverControllerName = "CertificateSigningRequestApproverController"
+)
+
+var (
+	errOrganizationNotAntrea    = fmt.Errorf("subject organization is not %s", antreaOrganization)
+	errDNSSANNotMatchCommonName = fmt.Errorf("DNS subjectAltNames do not match subject common name")
+	errEmailSANNotAllowed       = fmt.Errorf("email subjectAltNames are not allowed")
+	errIPSANNotAllowed          = fmt.Errorf("IP subjectAltNames are not allowed")
+	errURISANNotAllowed         = fmt.Errorf("URI subjectAltNames are not allowed")
+	errCommonNameRequired       = fmt.Errorf("subject common name is required")
+	errExtraFieldsRequired      = fmt.Errorf("extra values must contain %q and %q", sautil.PodNameKey, sautil.PodUIDKey)
+	errPodUIDMismatch           = fmt.Errorf("Pod UID does not match")
+	errPodNotOnNode             = fmt.Errorf("Pod is not on requested Node")
+)
+
+type approver interface {
+	recognize(csr *certificatesv1.CertificateSigningRequest) bool
+	verify(csr *certificatesv1.CertificateSigningRequest) error
+}
+
+// CSRApproverController is responsible for synchronizing the CertificateSigningRequest resources.
+type CSRApproverController struct {
+	client            clientset.Interface
+	csrLister         csrlister.CertificateSigningRequestLister
+	csrListerSynced   cache.InformerSynced
+	autoApprovePolicy config.AutoApprovePolicy
+	queue             workqueue.RateLimitingInterface
+	approvers         []approver
+}
+
+// NewCSRApproverController returns a new *CSRApproverController.
+func NewCSRApproverController(client clientset.Interface, csrInformer csrinformer.CertificateSigningRequestInformer, approvePolicy config.AutoApprovePolicy) *CSRApproverController {
+	c := &CSRApproverController{
+		client:            client,
+		csrLister:         csrInformer.Lister(),
+		csrListerSynced:   csrInformer.Informer().HasSynced,
+		autoApprovePolicy: approvePolicy,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+		approvers: []approver{
+			&ipsecCertificateApprover{
+				client:        client,
+				approvePolicy: approvePolicy,
+			},
+		},
+	}
+	csrInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueCertificateSigningRequest,
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueueCertificateSigningRequest(cur)
+			},
+			DeleteFunc: c.enqueueCertificateSigningRequest,
```

Comment:
Does it need to care update and delete events?

Reply:
removed.